chain/client/src/rocksdb_metrics.rs

```diff
@@ -0,0 +1,162 @@
+    ) -> anyhow::Result<()> {
+        for (stat_name, values) in stats {
+            if values.len() == 1 {
+                // A counter stats.
+                // A statistic 'a.b.c' creates the following prometheus metric:
+                // - near_a_b_c
+                if let StatsValue::Count(value) = values[0] {
+                    self.set_int_value(
+                        |stat_name| stat_name.to_string(),
+                        |stat_name| get_prometheus_metric_name(stat_name),
+                        stat_name,
+                        value,
+                    )?;
+                }
+            } else {
+                // A summary stats.
+                // A statistic 'a.b.c' creates the following prometheus metrics:
+                // - near_a_b_c_sum
+                // - near_a_b_c_count
+                // - near_a_b_c{quantile="0.95"}
+                for stats_value in values {
+                    match stats_value {
+                        StatsValue::Count(value) => {
+                            self.set_int_value(
+                                |stat_name| get_stats_summary_count_key(stat_name),
+                                |stat_name| get_metric_name_summary_count_gauge(stat_name),
+                                stat_name,
+                                *value,
+                            )?;
+                        }
+                        StatsValue::Sum(value) => {
+                            self.set_int_value(
+                                |stat_name| get_stats_summary_sum_key(stat_name),
+                                |stat_name| get_metric_name_summary_sum_gauge(stat_name),
+                                stat_name,
+                                *value,
+                            )?;
+                        }
+                        StatsValue::Percentile(percentile, value) => {
+                            let key: &str = stat_name;
+                            if !self.gauges.contains_key(key) {
+                                self.gauges.insert(
+                                    key.to_string(),
+                                    try_create_gauge_vec(
+                                        &get_prometheus_metric_name(stat_name),
+                                        stat_name,
+                                        &["quantile"],
+                                    )?,
+                                );
+                            }
+                            self.gauges
+                                .get(key)
+                                .unwrap()
+                                .with_label_values(&[&format!("{:.2}", *percentile as f64 * 0.01)])
+                                .set(*value);
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn set_int_value<F1, F2>(
+        &mut self,
+        key_fn: F1,
+        metric_fn: F2,
```

Comment:
Generic lambdas are a “death by a thousand cuts” way to increase compile times due to monomorphisation. In this case, we can use monomorphic fn pointer instead:

metric_fn: fn(&str) -> String

Reply:
Done